### ISSUE TITLE
Rule Proposal: Forbid DOM Nesting of Specific Elements

### ISSUE THREAD

**Issue by karlhorky** [yannickcr/eslint-plugin-react#3007]
Similar to [`react/forbid-elements`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/forbid-elements.md), it would make sense in certain circumstances to have the ability to prevent the nesting of certain DOM elements.

Eg. to prevent `<button>` from being used directly inside of `<a>` (or vice versa), which is a common pattern that we see with beginners:

```json
{
  "react/forbid-dom-nesting": [
    "error",
    {
      "rules": [
        "a > button",
        "button > a"
      ]
    }
  ]
}
```

Example of **incorrect** code:

```jsx
function LinkButton () {
  return (
    <a href="#1"><button>whoa</button></a>
  );
}
```

Examples of **correct** code:

```jsx
function LinkButton () {
  return (
    <a href="#1">whoa</a>
  );
}
```

---

To make it search through multiple levels of nesting, then use the space combinator instead of the child combinator:

```json
{
  "react/forbid-dom-nesting": [
    "error",
    {
      "rules": [
        "a button",
        "button a"
      ]
    }
  ]
}
```

For performance reasons, with the space combinator, maybe it would make sense to limit this by default to 5 or 10 levels of nesting.

**Comment by ljharb** [comment]
This would only be able to prevent same-component nesting, so I’m not sure how valuable it would be.
